usersync/cookie.go

```diff
@@ -20,70 +19,152 @@ const uidTTL = 14 * 24 * time.Hour
 
 // Cookie is the cookie used in Prebid Server.
 //
-// To get an instance of this from a request, use ParseCookieFromRequest.
-// To write an instance onto a response, use SetCookieOnResponse.
+// To get an instance of this from a request, use ReadCookie.
+// To write an instance onto a response, use WriteCookie.
 type Cookie struct {
-	uids   map[string]uidWithExpiry
+	uids   map[string]UIDEntry
 	optOut bool
 }
 
-// uidWithExpiry bundles the UID with an Expiration date.
-// After the expiration, the UID is no longer valid.
-type uidWithExpiry struct {
+// UIDEntry bundles the UID with an Expiration date.
+type UIDEntry struct {
 	// UID is the ID given to a user by a particular bidder
 	UID string `json:"uid"`
 	// Expires is the time at which this UID should no longer apply.
 	Expires time.Time `json:"expires"`
 }
 
-// ParseCookieFromRequest parses the UserSyncMap from an HTTP Request.
-func ParseCookieFromRequest(r *http.Request, cookie *config.HostCookie) *Cookie {
-	if cookie.OptOutCookie.Name != "" {
-		optOutCookie, err1 := r.Cookie(cookie.OptOutCookie.Name)
-		if err1 == nil && optOutCookie.Value == cookie.OptOutCookie.Value {
-			pc := NewCookie()
-			pc.SetOptOut(true)
-			return pc
-		}
+// NewCookie returns a new empty cookie.
+func NewCookie() *Cookie {
+	return &Cookie{
+		uids: make(map[string]UIDEntry),
 	}
-	var parsed *Cookie
-	uidCookie, err2 := r.Cookie(uidCookieName)
-	if err2 == nil {
-		parsed = ParseCookie(uidCookie)
-	} else {
-		parsed = NewCookie()
+}
+
+// ReadCookie reads the cookie from the request
+func ReadCookie(r *http.Request, decoder Decoder, host *config.HostCookie) *Cookie {
+	if hostOptOutCookie := checkHostCookieOptOut(r, host); hostOptOutCookie != nil {
+		return hostOptOutCookie
 	}
-	// Fixes #582
-	if uid, _, _ := parsed.GetUID(cookie.Family); uid == "" && cookie.CookieName != "" {
-		if hostCookie, err := r.Cookie(cookie.CookieName); err == nil {
-			parsed.TrySync(cookie.Family, hostCookie.Value)
-		}
+
+	// Read cookie from request
+	cookieFromRequest, err := r.Cookie(uidCookieName)
+	if err != nil {
+		return NewCookie()
 	}
-	return parsed
+	decodedCookie := decoder.Decode(cookieFromRequest.Value)
+
+	return decodedCookie
 }
 
-// ParseCookie parses the UserSync cookie from a raw HTTP cookie.
-func ParseCookie(httpCookie *http.Cookie) *Cookie {
-	jsonValue, err := base64.URLEncoding.DecodeString(httpCookie.Value)
+// PrepareCookieForWrite ejects UIDs as long as the cookie is too full
+func (cookie *Cookie) PrepareCookieForWrite(cfg *config.HostCookie, ttl time.Duration, encoder Base64Encoder) (string, error) {
+	encodedCookie, err := encoder.Encode(cookie)
 	if err != nil {
-		// corrupted cookie; we should reset
-		return NewCookie()
+		return "", err
 	}
 
-	var cookie Cookie
-	if err = json.Unmarshal(jsonValue, &cookie); err != nil {
-		// corrupted cookie; we should reset
-		return NewCookie()
+	isCookieTooBig := len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
```

Comment:
The validation for `MaxCookieSizeBytes`, just checks if the value is greater than `MIN_COOKIE_SIZE_BYTES`, but there is no non-negative check, so that's why I have this here.

The reason that this check is not present later in the function, is once it's checked once, I no longer need to check it again since the value never changes.